### ISSUE TITLE
Custom tab completion

### ISSUE THREAD

**Issue by Xandaros** [Blightmud/Blightmud#335]
It would be nice if it were possible to override the default completion behaviour.

I'm thinking of something like this:
`blight.on_tab_complete(callback: fun(current_input: string): string | nil)`

Multiple handlers may be registered. Execution order is undefined. The first handler that returns a string will cause that string to be appended to the input and prevent further handlers from being executed. Handlers who do not wish to handle completion should return `nil`. An empty string is a valid completion.

This could be useful in many cases:
- Commands could be auto-completed, like `/qui<tab>` turns into `/quit` or `/reconn` turning into `/reconnect`
- Command arguments (Full `/lua` completion, anyone? Probably not. But maybe arguments to `/connect` and the like)
- Basically all of the above, but with mud-specific commands. Particularly if you are in a mode with limited commands (Like Starmourn's hacking, for example)

**Comment by LiquidityC** [comment]
Sounds like a reasonable improvement. Especially if people want to make mud specific plugins.

**Comment by LiquidityC** [comment]
I don't recall if the current completion returns a full word or just the completion suffix. Either way this addition will have to match that behavior.

**Comment by Xandaros** [comment]
Yeah, I forgot to mention that the current completion behaviour will run as fallback if all handlers returned `nil`.

Though, to be fair, that is pretty obvious :P

**Comment by LiquidityC** [comment]
I took a look at this. The existing completion operates by returning a list of completions. Consecutive `<tab>` presses will step through this returned list without calling the completion function again. Most typing actions will clear the list and future `<tab>` presses will call the completion function again.

The custom completions will have to mimic this behavior by returning lists of completions. This would allow multiple custom completions to work in tandem by just concatenating the lists of completions that they return including the core implementation.

The order of the available completions will be based on the order of completion execution and the only guarantee we give is that the core completions will be last in that list.

To me this sounds like a better implementation. More in line with how eg. `bash_completion` works.

The one downside would be if a user has lots of custom completers which could potentially yield a very long list of completion options. But I don't really see this as a major problem. Same thing happens in most completion implementations if you start hitting tab after just typing one character.

Thoughts @Xandaros ?
